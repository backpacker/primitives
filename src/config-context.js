import React, { useContext, useMemo, useState } from 'react';
import PropTypes from 'prop-types';

import baseConfig from './config';

const ConfigContext = React.createContext(undefined);

function ConfigProvider({ config: customConfig, defaultTheme, children }) {
  const [activeTheme, setActiveTheme] = useState(defaultTheme);

  const theme = useMemo(() => {
    const config = { ...baseConfig, ...customConfig };

    const baseTheme = baseConfig.default;

    const defaultSpacerSize =
      config[activeTheme]?.defaultSpacerSize || baseTheme.defaultSpacerSize;
    const spacerUnit = config[activeTheme]?.spacerUnit || baseTheme.spacerUnit;

    const spacing = (size = defaultSpacerSize) => size * spacerUnit;

    return {
      ...config[activeTheme],
      spacing
    };
  }, [activeTheme, customConfig]);

  const value = {
    theme,
    setTheme: setActiveTheme
  };

  return (
    <ConfigContext.Provider value={value}>{children}</ConfigContext.Provider>
  );
}

ConfigProvider.propTypes = {
  config: PropTypes.shape({}),
  defaultTheme: PropTypes.string
};

ConfigProvider.defaultProps = {
  config: {},
  defaultTheme: 'default'
};

const withTheme = (Component) => (props) =>
  (
    <ConfigContext.Consumer>
      {(value) => <Component {...props} {...value} />}
    </ConfigContext.Consumer>
  );

function useTheme() {
  const ctx = useContext(ConfigContext);
  if (!ctx) {
    throw new Error(`useTheme() must be called within ConfigContext`);
  }
  return ctx;
}

export { ConfigProvider, withTheme, useTheme };
export default ConfigContext;
