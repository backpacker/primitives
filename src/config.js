export const weights = {
  extralight: '100',
  thin: '200',
  light: '300',
  normal: '400',
  medium: '500',
  semibold: '600',
  bold: '700',
  heavy: '800',
  extraheavy: '900'
};

const defaultTheme = {
  fontSize: {
    h1: 32,
    h2: 24,
    h3: 20,
    h4: 16,
    h5: 14,
    h6: 12,
    h7: 10,
    h8: 8
  },
  fontWeight: weights.normal,
  fontFamily: undefined,
  weights,

  spacerUnit: 8,
  defaultSpacerSize: 2,
  spacer: {
    xs: 10,
    s: 14,
    m: 18,
    l: 22,
    xl: 26
  },

  colors: {
    brand: '#28D9AA',
    background: '#FFF',
    text: '#000'
  },

  isDark: false
};

const darkTheme = {
  ...defaultTheme,
  colors: {
    brand: '#28D9AA',
    background: '#000',
    text: '#FFF'
  },
  isDark: true
};

export { defaultTheme };

export default {
  activeTheme: 'default',
  default: defaultTheme,
  dark: darkTheme
};
