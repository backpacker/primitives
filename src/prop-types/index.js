/**
 * Style PropTypes
 */

import LayoutStylePropTypes from './layout-style';
import TouchableStylePropTypes from './touchable-style';
import TextStylePropTypes from './text-style';
import ModalPropTypes from './modal';

export {
  LayoutStylePropTypes,
  TouchableStylePropTypes,
  TextStylePropTypes,
  ModalPropTypes
};
