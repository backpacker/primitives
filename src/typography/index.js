import React, { useContext } from 'react';

import Text from './base';
import ConfigContext from '../config-context';
import config from '../config';

const makeHeadingComponent = ({ key }) => {
  return ({ children, ...rest }) => {
    const { theme } = useContext(ConfigContext);
    const { fontSize } = theme;

    return (
      <Text fontSize={fontSize[key]} {...rest}>
        {children}
      </Text>
    );
  };
};

const [H1, H2, H3, H4, H5, H6, H7, H8] = Object.entries(
  config.default.fontSize
).map(([key]) => makeHeadingComponent({ key }));

export { Text, H1, H2, H3, H4, H5, H6, H7, H8 };
