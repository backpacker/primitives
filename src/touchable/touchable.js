import React from 'react';
import { TouchableOpacity } from 'react-native';
import PropTypes from 'prop-types';

import { TouchableStylePropTypes } from '../prop-types';
import { makeStyle } from '../utils/style';

const Touchable = (props) => {
  const { children, onPress, isDisabled, activeOpacity, ...rest } = props;

  const baseStyle = makeStyle([{ ...rest }]);

  return (
    <TouchableOpacity
      onPress={onPress}
      activeOpacity={activeOpacity}
      disabled={isDisabled}
      style={baseStyle}>
      {children}
    </TouchableOpacity>
  );
};

Touchable.propTypes = {
  ...TouchableStylePropTypes,
  onPress: PropTypes.func,
  isDisabled: PropTypes.bool,
  activeOpacity: PropTypes.number
};

Touchable.defaultProps = {
  onPress: () => {},
  isDisabled: false,
  activeOpacity: 0.5
};

export default Touchable;
