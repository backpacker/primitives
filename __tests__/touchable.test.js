/* eslint-env jest */
import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';

import { Touchable } from '../src/touchable';

describe('<Touchable>', () => {
  it('should render a touchable component', () => {
    const tree = <Touchable onPress={() => 42} testID='touchable' />;
    const { queryByTestId } = render(tree);
    const t = queryByTestId('touchable');
    const res = fireEvent.press(t);
    expect(res).toBe(42);
  });

  it('should render a disabled touchable component', () => {
    const tree = <Touchable onPress={() => 42} testID='touchable' isDisabled />;
    const { queryByTestId } = render(tree);
    const t = queryByTestId('touchable');
    const res = fireEvent.press(t);
    expect(res).toBeNull();
    expect(t).toBeDisabled();
  });
});
