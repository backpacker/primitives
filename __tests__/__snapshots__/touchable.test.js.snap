// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Touchable> should render a disabled touchable component 1`] = `
<TouchableOpacity
  activeOpacity={0.5}
  disabled={true}
  onPress={[Function]}
  style={Object {}}
/>
`;

exports[`<Touchable> should render a touchable component 1`] = `
<TouchableOpacity
  activeOpacity={0.5}
  disabled={false}
  onPress={[Function]}
  style={Object {}}
/>
`;
